### PR TITLE
[feature] API: support for UNIX sockets in client and server libraries

### LINKED ISSUE #121: API: support for binding server on unix socket

### PR DESCRIPTION
**Pull request by els0r**
Closes #121

### CODE REVIEW

File: pkg/api/client/client.go [hunk 1]
```diff
@@ -96,10 +104,30 @@ func NewDefault(addr string, opts ...Option) *DefaultClient {
 	for _, opt := range opts {
 		opt(c)
 	}
+
+	t := http.DefaultTransport
+
+	// change transport to dial to the unix socket instead
+	unixSocketFile := api.ExtractUnixSocket(addr)
+	unixIdent := "unix"
```
Comment (fako1024) [inline pkg/api/client/client.go:112]: Probably better to make a `const` out of this.

File: pkg/api/client/client.go [hunk 2]
```diff
@@ -96,10 +104,30 @@ func NewDefault(addr string, opts ...Option) *DefaultClient {
...
+		t = &http.Transport{
+			DialContext: func(ctx context.Context, _, _ string) (net.Conn, error) {
+				var d net.Dialer
+				return d.DialContext(ctx, unixIdent, unixSocketFile)
+			},
+		}
+		// also make sure to unset the address and modify the scheme so the calls to NewURL
+		// don't append the filename of the unix socket
+		c.hostAddr = ""
+		c.scheme = c.scheme + unixIdent
```
Comment (fako1024) [inline pkg/api/client/client.go:123]: Maybe also use something like `filepath.Join()` or `path.Join` to construct safe(r) URLs?

File: pkg/api/client/client.go [hunk 3]
```diff
@@ -189,6 +221,12 @@ func (c *DefaultClient) Modify(ctx context.Context, req *httpc.Request) *httpc.R
...
 
+// NewURL synthesizes a new URL for a given path depending on how the
+// client was configured.
+//
+// Example:
+//
+//	http://localhost:8145/status
 func (c *DefaultClient) NewURL(path string) string {
-	return fmt.Sprintf("%s://%s%s", c.scheme, c.hostAddr, path)
+	return fmt.Sprintf("%s%s%s", c.scheme, c.hostAddr, path)
```
Comment (fako1024) [inline pkg/api/client/client.go:231]: See comment above regarding `path.Join()` or similar.

**Review by fako1024 — APPROVED** [review]
LGTM, would be nice if you could have a quick look at the URL / path construction...

File: pkg/api/client/client.go (hunk 2 above)
Comment (els0r) [reply pkg/api/client/client.go:123]: Did that at first. What happened was that it turned `http://` into `http:/` 😦

**Pull request merged**